h/search/core.py

```diff
@@ -53,57 +65,74 @@ def run(self, params):
         :returns: The search results
         :rtype: SearchResult
         """
+
         total, annotation_ids, aggregations = self._search_annotations(params)
         reply_ids = self._search_replies(annotation_ids)
 
         return SearchResult(total, annotation_ids, reply_ids, aggregations)
 
     def clear(self):
         """Clear search filters, aggregators, and matchers."""
-        self.builder = query.Builder(es_version=self.es.version)
-        self.reply_builder = query.Builder(es_version=self.es.version)
-
-    def append_filter(self, filter_):
-        """Append a search filter to the annotation and reply query."""
-        self.builder.append_filter(filter_)
-        self.reply_builder.append_filter(filter_)
+        self._qualifiers = [query.Sorter()]
+        self._aggregations = []
 
-    def append_matcher(self, matcher):
-        """Append a search matcher to the annotation and reply query."""
-        self.builder.append_matcher(matcher)
-        self.reply_builder.append_matcher(matcher)
+    def append_qualifier(self, qualifier):
+        """Append a search filter, matcher, etc to the search query."""
+        # Note we must insert any new qualifier at the begining of the list
+        # since the KeyValueFilter must always be run after all the other
+        # qualifiers.
+        self._qualifiers.insert(0, qualifier)
 
     def append_aggregation(self, aggregation):
-        self.builder.append_aggregation(aggregation)
+        """Append an aggregation to the search query."""
+        self._aggregations.append(aggregation)
 
-    def _search_annotations(self, params):
-        if self.separate_replies:
-            self.builder.append_filter(query.TopLevelAnnotationsFilter())
+    def _search(self, qualifiers, aggregations, params):
+        """
+        Applies the qualifiers, aggregations, and executes the search.
+        """
+        search = elasticsearch_dsl.Search(
+            using="default", index=self.es_index
+        )
+
+        for agg in aggregations:
+            agg(search, params)
+        for qual in qualifiers:
+            search = qual(search, params)
 
         response = None
         with self._instrument():
-            response = self.es.conn.search(index=self.es.index,
-                                           doc_type=self.es.mapping_type,
-                                           _source=False,
-                                           body=self.builder.build(params))
+            response = search.execute()
```

Comment:
It looks like you can do something like this:
```
s = s.extra(_source=False)
```
which presumably could be another qualifier.

Reply:
The `h.search.core.Search` class only ever returns annotation IDs from the `run` method, so I think the `s.extra(_source=False)` step should go in the class directly rather than being in an external class.

As I see it, the only reason of having the qualifiers/modifiers be separate from the `Search` class is to be able to test their effects independently. We don't need to do that with `_source=False`.

Reply:
Were you wanting a test for this? I'm not sure there's a good way to do this as the code is written right now as it's all hidden within the _search method. I suppose we could mock it and make sure source(False) got called or something?

Reply:
If you did want to test this, I think mocking `elasticsearch_dsl.Search` would be the easiest thing to do. There is no logic involved here so I'm not too worried about it.